sky/adaptors/azure.py

```diff
@@ -55,6 +62,51 @@ def get_client(name: str, subscription_id: str):
         elif name == 'resource':
             from azure.mgmt.resource import ResourceManagementClient
             return ResourceManagementClient(credential, subscription_id)
+        elif name == 'storage':
+            from azure.mgmt.storage import StorageManagementClient
+            return StorageManagementClient(credential, subscription_id)
+        elif name == 'authorization':
+            from azure.mgmt.authorization import AuthorizationManagementClient
+            return AuthorizationManagementClient(credential, subscription_id)
+        elif name == 'graph':
+            from msgraph import GraphServiceClient
+            return GraphServiceClient(credential)
+        elif name == 'container':
+            from azure.storage.blob import ContainerClient
+            container_url = (f'https://{storage_account_name}.'
+                             f'blob.core.windows.net/{container_name}')
+            container_client = ContainerClient.from_container_url(container_url)
+            try:
+                container_client.exists()
+            except exceptions().ClientAuthenticationError as e:
+                # Caught when credential is not provided to the private
+                # container url or wrong container name is provided as the
+                # public container url. We reattempt with credentials assuming
+                # user is using private container with access rights.
+                if 'ErrorCode:NoAuthenticationInformation' in e.message:
+                    container_client = ContainerClient.from_container_url(
+                        container_url, credential)
```

Comment:
The reason why I had to call the api twice is because azure python sdk currently seems to lack feature to determine either the given container url is public or private. I did some extensive research to see if there's a method for our use case to do this, but it didn't seem to have a working solution. So the hacky way I figured out was to run `container_client.exists()` without the credential passed to see if it's private or public. If it's private, it'll throw an error which confirms that it's private, and then, I call again with the credentials. So it is necessary to call the API twice in this case.

But just to make sure, I filed an [issue](https://github.com/Azure/azure-sdk-for-python/issues/35770) on azure sdk's repo asking if there's a good method for us to use. If they suggest a better way, we can go with that.

Reply:
Sounds great! Lets see the response to the issue ;)

Reply:
It's impossible to do so without try-except block just like what I did with my implementation: https://github.com/Azure/azure-sdk-for-python/issues/35770#issuecomment-2142815667

Reply:
good point! Left it in the comment.

Reply:
Lets keep the issue link in the code as well.

Reply:
Oh sry i missed it. IThe issue link is already included. nvm